client-core/src/service/wallet_service.rs

```diff
@@ -91,53 +127,16 @@ pub struct Wallet {
 impl Encode for Wallet {
     fn encode_to<W: Output>(&self, dest: &mut W) {
         self.view_key.encode_to(dest);
-        (self.key_pairs.len() as u64).encode_to(dest);
-        for (key, value) in self.key_pairs.iter() {
-            key.encode_to(dest);
-            value.encode_to(dest);
-        }
-        (self.public_keys.len() as u64).encode_to(dest);
-        for key in self.public_keys.iter() {
-            key.encode_to(dest);
-        }
-        (self.staking_keys.len() as u64).encode_to(dest);
-        for key in self.staking_keys.iter() {
-            key.encode_to(dest);
-        }
-        (self.root_hashes.len() as u64).encode_to(dest);
-        for hash in self.root_hashes.iter() {
-            hash.encode_to(dest);
-        }
     }
 }
 
 impl Decode for Wallet {
     fn decode<I: Input>(input: &mut I) -> std::result::Result<Self, parity_scale_codec::Error> {
         let view_key = PublicKey::decode(input)?;
-        let mut key_pairs = BTreeMap::new();
-        let len = u64::decode(input)?;
-        for _ in 0..len {
-            let key = PublicKey::decode(input)?;
-            let value = PrivateKey::decode(input)?;
-            key_pairs.insert(key, value);
-        }
-        let len = u64::decode(input)?;
-        let mut public_keys = IndexSet::with_capacity(len as usize);
-        for _ in 0..len {
-            public_keys.insert(PublicKey::decode(input)?);
-        }
-
-        let len = u64::decode(input)?;
-        let mut staking_keys = IndexSet::with_capacity(len as usize);
-        for _ in 0..len {
-            staking_keys.insert(PublicKey::decode(input)?);
-        }
-
-        let len = u64::decode(input)?;
-        let mut root_hashes = IndexSet::with_capacity(len as usize);
-        for _ in 0..len {
-            root_hashes.insert(H256::decode(input)?);
-        }
+        let key_pairs = BTreeMap::new();
+        let public_keys = IndexSet::new();
+        let staking_keys = IndexSet::new();
+        let root_hashes = IndexSet::new();
```

Comment:
Are these fields still necessary if they are read from storage directly?

Reply:
other sync code use it, so to be compatible with other sources

Reply:
@leejw51crypto do you have some example where these fields are used directly? 
In sync, I only see it uses `wallet.staking_addresses()`, for example, and this function was changed to load from storage instead of that field?

Reply:
If they are never used, they can perhaps be removed from the `struct Wallet` ?

Reply:
agree, i'll remove it after check

Reply:
but that's also taken directly from the storage: https://github.com/crypto-com/chain/pull/1278/files#diff-fcaa70a81e2fb374ccdf66db46aeae86R565 ?

Reply:
yes, that's possible, 
but need to change syncer_logic, sync.

Reply:
also memory stoage should use BTreeMap, need to apply after this pr

Reply:
@leejw51crypto no, my question / observation was that syncer_logic / sync seem to only call methods on `Wallet` -- and these methods were changed in this PR -- so it seems those fields in `Wallet` struct are never used, even in syncer_logic / sync ?

Reply:
Wallet syncer uses `wallet.view_key`/`wallet.staking_addresses()`/`wallet.transfer_addresses()`.

Reply:
will merge this PR; some improvements (removing these redundant fields; safer concurrency etc. can be in other PRs)

Reply:
ok